src/ngx_http_lua_resolve.c

```diff
@@ -0,0 +1,483 @@
+
+/*
+ * Copyright (C) Sergey Kharkhardin (slimboyfat)
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_resolve.h"
+#include "ngx_http_lua_contentby.h"
+
+
+typedef struct ngx_http_lua_resolve_ctx_s
+        ngx_http_lua_resolve_ctx_t;
+
+
+static int ngx_http_lua_ngx_resolve(lua_State *L);
+static void ngx_http_lua_resolve_handler(ngx_resolver_ctx_t *ctx);
+static void ngx_http_lua_resolve_cleanup(void *data);
+static void ngx_http_lua_resolve_empty_handler(ngx_resolver_ctx_t *ctx);
+static ngx_int_t ngx_http_lua_resolve_resume(ngx_http_request_t *r);
+static int ngx_http_lua_resolve_select_retval(ngx_http_lua_resolve_ctx_t *u,
+    lua_State *L);
+static int ngx_http_lua_resolve_get_retval(struct sockaddr *sockaddr,
+    socklen_t socklen, lua_State *L);
+static int ngx_http_lua_resolve_set_query_option(const char *name,
+    unsigned *option, lua_State *L);
+
+struct ngx_http_lua_resolve_ctx_s {
+    ngx_http_request_t            *request;
+    ngx_http_upstream_resolved_t  *resolved;
+    ngx_http_lua_co_ctx_t         *curr_co_ctx;
+    unsigned                      ipv4:1;
+    unsigned                      ipv6:1;
+};
+
+
+static int ngx_http_lua_ngx_resolve(lua_State *L)
+{
+    int                          n, saved_top;
+    size_t                       len;
+    u_char                      *p;
+    ngx_int_t                    rc;
+    ngx_str_t                    host;
+    ngx_url_t                    url;
+    ngx_http_lua_ctx_t          *ctx;
+    ngx_http_request_t          *r;
+    ngx_resolver_ctx_t          *rctx, temp;
+    ngx_http_lua_co_ctx_t       *coctx;
+    ngx_http_core_loc_conf_t    *clcf;
+    ngx_http_lua_resolve_ctx_t  *u;
+
+    unsigned  ipv4 = 1;
+    unsigned  ipv6 = 0;
+
+    n = lua_gettop(L);
+    if (n != 1 && n != 2) {
+        return luaL_error(L, "passed %d arguments, but accepted 1 or 2", n);
+    }
+
+    if (lua_type(L, 1) != LUA_TSTRING) {
+        return luaL_error(L, "1st parameter must be a string");
+    }
+
+    if (n == 2) {
+        if (lua_type(L, 2) != LUA_TTABLE) {
+            return luaL_error(L, "2nd parameter must be a table");
+        }
+
+        rc = ngx_http_lua_resolve_set_query_option("ipv4", &ipv4, L);
+        if (rc != NGX_OK) {
+            return rc;
+        }
+
+        rc = ngx_http_lua_resolve_set_query_option("ipv6", &ipv6, L);
+        if (rc != NGX_OK) {
+            return rc;
+        }
+    }
+
+    r = ngx_http_lua_get_req(L);
+    if (r == NULL) {
+        return luaL_error(L, "no request found");
+    }
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return luaL_error(L, "no ctx found");
+    }
+
+    ngx_http_lua_check_context(L, ctx, NGX_HTTP_LUA_CONTEXT_REWRITE
+                                       | NGX_HTTP_LUA_CONTEXT_ACCESS
+                                       | NGX_HTTP_LUA_CONTEXT_CONTENT);
```

Comment:
The “timer" and "ssl cert" contexts should work as well, right?

Reply:
I didn't cover these cases by tests, but, it seems it should work.